defender/staged/downloaded/Compound_Gasless_Voting.js

```diff
@@ -0,0 +1,95 @@
+const ethers = require('ethers');
+
+const { DefenderRelayProvider, DefenderRelaySigner } = require('defender-relay-client/lib/ethers');
+
+const compoundGovernanceAddress = '0xc0Da02939E1441F497fd74F78cE7Decb17B66529'; // GovernorBravoDelegate.sol
+
+const compTokenAddress = '0xc00e94Cb662C3520282E6f5717214004A7f26888'; // Comp.sol
+
+const compoundGovernanceAbi = [
+  'function castVoteBySig(uint proposalId, uint8 support, uint8 v, bytes32 r, bytes32 s)',
+];
+
+const compTokenAbi = [
+  'function delegateBySig(address delegatee, uint nonce, uint expiry, uint8 v, bytes32 r, bytes32 s)',
+];
+
+exports.handler = async function handler(autotaskEvent) {
+  if (autotaskEvent.request === undefined) {
+    throw new Error('POST request information not found');
+  }
+  // body contains batch of signed messages to submit on-chain through the Defender Relayer
+  // the api endpoint does most of the validation checking for us already
+  // (ex. checks for valid signatures and all required fields)
+  // https://app.swaggerhub.com/apis-docs/arr00/COMP.vote/1.0#/
+  const { body } = autotaskEvent.request;
+
+  if (Array.isArray(body) === false) {
+    throw new Error('Request body must be an Array');
+  }
+
+  // initiate a new relay provider
+  console.debug('Creating DefenderRelayProvider');
+  const provider = new DefenderRelayProvider(autotaskEvent);
+
+  // initiate a new relay signer
+  console.debug('Creating DefenderRelaySigner');
+  const signer = new DefenderRelaySigner(autotaskEvent, provider, { speed: 'fast' });
+
+  // create an ethers.js Contract Object to interact with the on-chain governance smart contract
+  console.debug('Creating governanceContract');
+  const governanceContract = new ethers.Contract(
+    compoundGovernanceAddress,
+    compoundGovernanceAbi,
+    signer,
+  );
+
+  // create an ethers.js Contract Object to interact with the on-chain COMP smart contract
+  console.debug('Creating compContract');
+  const compTokenContract = new ethers.Contract(
+    compTokenAddress,
+    compTokenAbi,
+    signer,
+  );
+
+  // cast vote or delegate vote on-chain
+  const promises = body.map(async (votes) => {
+    if (votes.delegatee !== undefined) {
+      const {
+        address, delegatee, nonce, expiry, v, r, s,
+      } = votes;
+
```

Comment:
Should we sanity check these values before spending the gas to make the call?

Reply:
I believe the API endpoint does the sanity checks for us https://app.swaggerhub.com/apis-docs/arr00/COMP.vote/1.0#/default/delegatePOST - if you try to send a post request to it with invalid fields it will won't accept it

Reply:
You can test that it does validate signatures by pasting this into the data field and executing a POST request to the API endpoint:
```
{
  "address": "0x1111111111111111111111111111111111111111",
  "delegatee": "0x2222222222222222222222222222222222222222",
  "nonce": 0,
  "expiry": 10,
  "v": "0x1b",
  "r": "0xda4429a9e8e6b54cb101b2df002039f2879ab4ca0e8fae64134942cb81f3e581",
  "s": "0x3b90a37dc078a82dfc418695b1d4473661aa4d24dd874ac68678894ff44a6b27"
}
```

It should give back an error messaging stating "given address does not match signer address"

Reply:
Added sanity check